### ISSUE TITLE
Jar file issue

### ISSUE THREAD

**Issue by wholder** [marytts/marytts#980]
I know this issue has come up before, but none of the solutions I've read about here seem to work for me.  I'm working on a game project where I'd like to use MaryTTS, so I built a simple test project in InteliJ and managed to get it working fairly easily.  For simplicity, I only used the following JAR files from the 5.2 release build;
```
  marytts-lang-en-5.2.jar
  marytts-runtime-5.2-jar-with-dependencies.jar
  voice-cmu-rms-hsmm-5.2.jar

```
and I invoke MaryTTS like this:
```
  MaryInterface marytts = new LocalMaryInterface();
  marytts.setVoice("cmu-rms-hsmm");
  marytts.setAudioEffects("JetPilot+F0Add(f0Add:-20.0;)");
  say("Hello World!");         // calls marytts.generateAudio()
```
This all works fine when I run, or debug inside IntelliJ.  However, when I try to build a runnable JAR file and then run it, like this:

`
  java -jar MaryTTSDemo.jar
`

It prints out an error, like this:
```
  marytts.exceptions.MaryConfigurationException: Cannot start MARY server
	at marytts.LocalMaryInterface.<init>(LocalMaryInterface.java:66)
	at MaryDemo.<init>(MaryDemo.java:53)
	at MaryDemo.main(MaryDemo.java:185)
  Caused by: java.lang.NullPointerException
	at marytts.util.MaryRuntimeUtils.ensureMaryStarted(MaryRuntimeUtils.java:70)
	at marytts.LocalMaryInterface.<init>(LocalMaryInterface.java:64)
	... 2 more
```

When I inspect the .jar file created by IntelliJ, it looks like this:

![Screen Shot 2020-05-05 at 3 39 03 PM](https://user-images.githubusercontent.com/1062514/81122823-9d3f1200-8ee6-11ea-8183-a90fa7001f2c.png)

and seems to contain all the elements I would expect to find.  From what I've read in posts about other, similar issues, this is due to some type of problem with the way MaryTTS' "service loaders", which seems to be some type of static initialization mechanism in the `ServiceLoader` class.  However, so far, I have been unable to find a way to build a workable .jar file in IntelliJ.

I think MaryTTS has real potential for the project I'm working, so I plan to try and get more  familiar with it and perhaps find  ways to trim down its size a bit and add some features that would be useful in a game application.  But, so far, the overall structure seems, to me, rather complex and hard to debug, as so many important things seem to happen as side effects of class loading.  

For example, if I set a Java Field Watchpoint on this line in the `MaryConfig` class:

`  private static final ServiceLoader<MaryConfig> configLoader = ServiceLoader.load(MaryConfig.class);
`
And then run the project normally in IntelliJ (not as a packaged jar file), the code will stop at the Watchpoint, as expected.  But, when I tell IntelliJ to resume, the program pauses for a very long time (36 seconds, or more) before it speaks (much longer than the normal, 2-3 second delay when it starts up without the Watchpoint set.) Very odd behavior, IMO.

Wayne

**Comment by psibre** [comment]
Thanks for the detailed feedback!

The issue here regarding the service loaders is that normally each MaryTTS component is packaged in its own Jar file with a single service loader. It is located here:
```
META-INF/services/marytts.config.MaryConfig
```
and its content is a single line pointing to the component config to load, e.g.,
```
marytts.voice.CmuSlt.Config
```
for the female English `voice-cmu-slt` unitselection voice.

If multiple MaryTTS components are combined inside a single Jar, the service loaders need to be *merged* -- otherwise MaryTTS will not be able to load all of the components when it starts. This is more or less trivial once you realize that this simply means the service loader file needs to include one line for *each* component. E.g., to package `voice-cmu-slt` with the English language component and the main MaryTTS runtime in a single Jar, the service loader would contain this:
```
marytts.voice.CmuSlt.Config
marytts.config.MainConfig
marytts.language.en.EnglishConfig
```
What's more, this sort of transformation is provided by build plugins, such as [https://plugins.gradle.org/plugin/com.github.johnrengelman.plugin-shadow](com.github.johnrengelman.plugin-shadow), so all you need to do is
```gradle
shadowJar {
  mergeServiceFiles()
}
```
See https://github.com/marytts/marytts-txt2wav and duplicate issues #873,  #725, #690, etc.

**Comment by wholder** [comment]
Thanks for taking the time to reply and for trying to help.

First thing, I tried to follow the first link inn your reply, but it says "Not found" Did you mean to type the link:  [https://github.com/johnrengelman/shadow](https://github.com/johnrengelman/shadow)?

So far I've just bee using the prebuilt JAR files for the 5.2 build which I downloaded from:

  [http://mary.dfki.de/download/index.html](http://mary.dfki.de/download/index.html)

Are you saying that I need to download the full GitHub project and reconfigure it to build the JAR files in a different way?  Or, is the shadow tool you referenced able to modify the 3 JAR files I'm using to combine the service loaders?  Frankly, I'm not that familiar with gradle-based build processes, so please forgive me for not understand exactly what you're trying to explain to me.

I have also not used the ServiceManager class before, but I now see that the META-INF folder in each jar file, such as `marytts-lang-en-5.2.jar` contains a folder named "`services`" that looks like this:

![services](https://user-images.githubusercontent.com/1062514/81221875-47c33d80-8f98-11ea-938a-dcc98d82778c.png)

where the contents of the `marytts.config.MaryConfig ` file in the "services" folder contains:

`marytts.language.en.EnglishConfig`

and in the `marytts-runtime-5.2-jar-with-dependencies.jar` file, I see this:

![services3](https://user-images.githubusercontent.com/1062514/81222260-da63dc80-8f98-11ea-9ed2-5e87f1545a05.png)

where the contents of the `marytts.config.MaryConfig` file in the "services" folder contains:

`marytts.config.MainConfig`

and in the `voice-cmu-rms-hsmm-5.2.jar` file I see:

![services4](https://user-images.githubusercontent.com/1062514/81222378-0c753e80-8f99-11ea-9baf-f0ecc9003363.png)

where the contents of the `marytts.config.MaryConfig ` file in the "services" folder contains:

`marytts.voice.CmuRmsHsmm.Config`

However, when I look in the combined JAR file that IntellJ created, I see the contexts of the "services" folder looks like this:

![services2](https://user-images.githubusercontent.com/1062514/81222036-8527cb00-8f98-11ea-8471-72bfa485b37f.png)

and the content of the `marytts.config.MaryConfig ` file in the "services" folder contains only a single line:

`marytts.language.en.EnglishConfig`

So, I'm wondering if the issue is that IntelliJ is not handling the build process properly and is not combining the contents of identically-named files in the "services" folder into a single file in the output JAR file, as it seems like doing that would solve the problem.  Or, am I misunderstand the issue?

Wayne

**Comment by psibre** [comment]
> First thing, I tried to follow the first link inn your reply, but it says "Not found" Did you mean to type the link: https://github.com/johnrengelman/shadow?

Yes, sorry. Fixed the link.

**Comment by psibre** [comment]
Regarding the *usage* of the shadow plugin with Gradle, there are various ways you could get this into your project. If you're using Maven, there is also the [Shade Plugin](http://maven.apache.org/plugins/maven-shade-plugin), which has a [ServicesResourceTransformer](http://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ServicesResourceTransformer) for this purpose.

Having said that, let me mention that personally, I believe that a Java project -- whether it is built with Gradle or Maven -- should have all build logic configured by the build system, so that there is a *single source of truth* for how the project is built, and it can be done just as well from the command line or in an CI environment. An IDE should use *only* this build logic when it imports the project, and adding random build customization via the GUI is only asking for trouble. (Incidentally, this is one of the reasons why IDE project metadata files should not be committed to, but ignored by, SCM).

Fortunately, IntelliJ IDEA is extremely good at importing Gradle (and Maven) projects, which is the main reason I switched away from Eclipse so many years ago; I was sick and tired of having to install and update so many plugins and jump through endless dialog windows, just to be able to work with Gradle projects in Eclipse. Of course I could be using Gradle's tooling plugins to generate correct project files for various IDEs, but with IDEA everything just works right out of the box! =)

TL;DR: I'm not sure how your IDEA project is generating the combined Jar without seeing your build logic. If you have your project source code on GitHub (or elsewhere online), we can help you get the configuration right.

**Comment by wholder** [comment]
I've posted my IntelliJ test project at: [https://github.com/wholder/MaryDemo](https://github.com/wholder/MaryDemo)

I consists of the three jar files extracted from the "lib" folder in a Zip file named `marytts-5.2.zip` which was downloaded from the [MaryTTS 5.2 download page](http://mary.dfki.de/download/index.html).  To this I added a simple test class I wrote called `MaryDemo.java`.  Nothing fancy.  For reference I'm running the latest version of IntelliJ Community 2020.1 on a MacPro running macOS Mojave 10.14.6, 

Also, to try a different approach, I used IntelliJ to download the [MaryTTS project](https://github.com/marytts/marytts) from GitHub and then let it build and then ran the "Build Artifacts" process.  This produced a JAR file named `marytts.jar` in the `out/artifacts/marytts_jar` folder.  I then tried to use this JAR file (and `voice-cmu-rms-hsmm-5.2.jar`) in a new project I named `MaryDemo2` to which I added the same `MaryDemo.java` test class, but got this exception when I tried to run my test code in Intellij (I did not bother to build a .jar file from this project):

```
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.NoClassDefFoundError: marytts/MaryInterface
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.privateGetMethodRecursive(Class.java:3048)
	at java.lang.Class.getMethod0(Class.java:3018)
	at java.lang.Class.getMethod(Class.java:1784)
	at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
Caused by: java.lang.ClassNotFoundException: marytts.MaryInterface
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 7 more

```
I've  uploaded this IntelliJ project to Github as [MaryDemo2](https://github.com/wholder/MaryDemo2).

As to using an automated build process, I don't disagree.  However, when I'm first starting to get familiar with a new library, I like to keeps things as simple as possible.  To create both the `MaryDemo` and `MaryDemo2` projects I tried to let IntelliJ handle the work.  Here's the process I followed:

1. Create new Java project in IntelliJ.
2. Create "lib" folder in project folder and copy in the MaryTTS JAR files.
3. Open "Project Structure" panel, select "Libraries", click the "+" sign then select "Java" and then select and add the JAR files in the lib folder.
4. Copy MaryDemo.java into the "src" folder.
5. Run Build->Build Project.
6. Right click on "MaryDemo" in the src folder in the Intellij project pane and select Run 'MaryDemo.main()' to run the code.

To create the output JAR file for the MaryDemo Github project, I used this process:

1. Open "Project Structure" panel, select "Artifacts", click the "+" sign then select "JAR" and then select "From module with dependencies", then select "MaryDemo" as the main class, then click OK and OK again.
2. Run Build->Build Artifacts then click "Build"

That's pretty much it.

Wayne

**Comment by psibre** [comment]
OK, thanks for the link to your project code. Unless you accidentally omitted lots of critical files, it looks like you're not using any build system *at all*. If you're creating the entire project only by clicking stuff in IntelliJ IDEA, and there are no Maven or Gradle (or even Ant) project files to define and configure your project, then all bets are off. There's no way I can help if the project is not reproducible by defining the build logic in any of several established ways.

I strongly suggest you adopt Maven or Gradle as your build system. Sure, if you've never worked with something like that before, there's a learning curve, but it's worth it. It's also the only sensible way to integrate MaryTTS, and the only way we can provide support.

See the official guides, e.g.,
- https://guides.gradle.org/creating-new-gradle-builds/
- https://guides.gradle.org/building-java-libraries/
- https://guides.gradle.org/building-java-applications/
- https://guides.gradle.org/consuming-jvm-libraries/

Also countless tutorials and blog posts around the web.

**Comment by wholder** [comment]
After a bit of experimentation, I did find a simple solution to my initial problem by using the jar utility to insert a new `marytts.config.MaryConfig` file into the `META-INF/services` folder in my `MaryTTSDemo.jar file`.  This file contains the following lines:

```
  marytts.language.en.EnglishConfig
  marytts.voice.CmuRmsHsmm.Config
  marytts.config.MainConfig
```

In addition, I contacted the support staff for Intellij and inquired whether the IDE's failure to handle this was a bug.  Just today I received this reply from Nikolay Chashnikov:

_"You're right, IDE should be able to handle this automatically, but currently it isn't implemented. You may keep separate JAR files and add them to the classpath of your JAR via Class-Path attribute in MANIFEST.MF file."_

Hopefully, the folks at Intellij will eventually address this.  For time being, I may try Nikolay's suggestion, or I may just use ANT to automate my jar-patching solution.  However, I would really like to find a way to greatly pare down the MarytTTS code to just the set of features that I will need for my project.

Also, out of curiosity, I ran the Intellij Analysis tool on the `marytts` project and noticed quite a few areas where the code could potentially be modernized, improved, or simplified.  Is there a plan to address any of these issues in the near future?

Wayne

**Comment by psibre** [comment]
Great to hear you solved the issue by creating a "merged" service loader file (by hand, evidently).

Interesting to hear that the JetBrains developers view this as missing functionality. But like I said, I would normally expect this to be handled by the build tool, and I suspect the IDE tooling should then be able to handle the assembled component automagically.

I guess we can close this issue then.

**Comment by psibre** [comment]
> Also, out of curiosity, I ran the Intellij Analysis tool on the marytts project and noticed quite a few areas where the code could potentially be modernized, improved, or simplified. Is there a plan to address any of these issues in the near future?

Some of the code dates back to the early noughties. We're doing our best to clean it up, but since we're not full-time developers, it takes time. See however PRs like #977 or #968 for recent progress.

**Comment by wholder** [comment]
I decided to investigate the various Gradle scripts in the project and got some unexpected errors.  For example, when I run this script:

![Screen Shot 2020-05-09 at 10 45 47 PM](https://user-images.githubusercontent.com/1062514/81491781-dbba3100-9246-11ea-8a5c-594f6dd0c894.png)

in macOS IntelliJ, I get this odd error:

```
/Users/wholder/IdeaProjects/marytts/marytts-runtime/src/main/java/marytts/signalproc/effects/BaseAudioEffect.java:30: error: unmappable character for encoding ASCII
// It??s main functionality is providing base functions for parsing parameter strings
     ^
```
To make sure this wasn't something my other experiment corrupted, I trashed my original project and fetched it again fresh from Githib.

Similarly, if I run this Gradle script:

![Screen Shot 2020-05-09 at 10 49 10 PM](https://user-images.githubusercontent.com/1062514/81491811-5420f200-9247-11ea-8f5e-f0e211c4821d.png)

I get this error:

```
/Users/wholder/IdeaProjects/marytts/marytts-common/src/main/java/marytts/util/MaryUtils.java:129: error: unmappable character for encoding ASCII
		final String letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ??????abcdefghijklmnopqrstuvwxyz????????";
		                                                  ^
```
I'm not sure what's going on, as the files look fine in IntelliJ's editor. Do you have any idea what I'm doing wrong?

Wayne

**Comment by psibre** [comment]
The source encoding in your IDE project is wrong.
Our Gradle build logic sets this:
```gradle
tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
}
```
If you *imported* the Gradle project as opposed to manually configuring it (which evidently bypasses the Gradle build logic), you would not face these issues.

**Comment by wholder** [comment]
I used IntelliJ to import the project.  And, as I mentioned the files look correct when I open them in intellij’s editor.  That is, the correct Unicode characters are shown.  It’s only the error message from the Gradle script where the incorrect characters are shown.  Is there perhaps a setting for the Gradle plug in I need to change?

Wayne

**Comment by psibre** [comment]
Sorry, my bad!

You're right, there's definitely something awry with how the latest IntelliJ IDEA import the current MaryTTS master (b4c8a1c09f761bdb8bd4dfadfc55158a2ceb9f32).

Importing in IDEA 2019.3.1 running under Mojave on my MacBook Pro is fine. But after upgrading IDEA to 2020.1.1 under Catalina on my Mac Pro at the office I do in fact now see errors upon import. Seems like something is wrong with the Gradle tooling in the latest version of IDEA.

Here's a workaround:
```diff
diff --git a/build.gradle b/build.gradle
index daff36756..ce9d63364 100644
--- a/build.gradle
+++ b/build.gradle
@@ -3,6 +3,8 @@ plugins {
 }

 allprojects {
+    apply plugin: 'idea'
+
     group 'de.dfki.mary'
     version '5.3-SNAPSHOT'
```
This allows me to run `./gradlew idea`, and then successfully open the MaryTTS project via the generated `marytts.ipr` -- minus the tooling in IDEA... But various other functionality is broken as well -- I can build the project only from the Terminal, but not in IDEA 2020.

We're currently heavily restructuring the project in preparation for a new release, and I'll pay special attention to the ability to import the project in IDEA 2020 without errors.

**Comment by wholder** [comment]
Whew... I'm glad to hear it wasn't something I screwed up.  Also, while I'm using IDEA 2020.1.1, I'm still running on macOS Mojave (still have some important 32 bit apps I need to use.)

it's interesting to hear you're working on restructuring the project, as that is what I was planning to work on next.  I'm sure you have other priorities, but, it would be nice to be able to easily separate and remove code that's not related to direct synthesis of speech without having to start and maintain a server running in the background.

BTW,  it may take me some time to get up to speed, but I'm hoping to find ways I can contribute to this project once I get a better idea of how it all hangs together.

Wayne

**Comment by psibre** [comment]
Thanks for the feedback!

Incidentally, I'm not seeing any problems with IDEA 2020.1.1 on Ubuntu Bionic... Maybe my issue is unrelated to this one, and more a problem with Catalina than IDEA.

**Comment by wholder** [comment]
I’m using MacOS Mojave.  Is there a specific test I can try to help narrow this down?

Wayne
